### ISSUE TITLE
Feature request: Plot graph using graphviz

### ISSUE THREAD

**Issue by pinpox** [mickael-menu/zk#48]
It would be nice to be able to visualise the connections between notes as a graph. I have seen this functionality on other applications implementing that implement a Zettelkasten-style notebook and find it very helpful. Here is an [example of how that could look](https://github.com/MaxvdKolk/vimwikigraph#examples). 

I propose a command,  e.g. `zk graph` that, that just outputs the source for a graph as a graphviz graph, allowing to pipe it to a file or render it using `dot`.

Here is [another example](https://github.com/joashxu/zetteltools) of a very useful visualisation, haven't been able to run it on my `zk` notes folder though.

![image](https://user-images.githubusercontent.com/1719781/122354965-c63b2e80-cf51-11eb-99ae-e76545618737.png)

Maybe there already is a solution for this I'm not aware of, if so please let me know!

**Comment by mickael-menu** [comment]
I had something similar in mind 👍 

I thought about having something like `zk graph --format <format> [filtering]`, using the same filtering options as `list` to get a subset of the graph.

Format could be `json`, `dot`, `svg` and more. And `ascii` format would be great, but that might stretch the terminal a bit too much!

An `--interactive` (xor with `--format`) mode serving an HTML version on the browser allowing to filter interactively the graph would be nice as well.

Initially I wanted to produce a JSON version, because it might be easier to pipe through other programs, including D3.js (which the project you shared uses). But I'm still trying to figure out the best JSON schema. There's a few spec out there but unfortunately nothing really standard:
* https://netflix.github.io/falcor/documentation/jsongraph.html
* http://jsongraphformat.info/
* JSON-LD too, but it might be a bit too complex for this use case

Making a prototype in DOT could be a first step. There's [even a Go package](https://github.com/emicklei/dot) to generate DOT graphs. I'm just wondering if DOT is really appropriate when you have so many nodes. But it could be useful for a small subset of your ZK.

**Comment by pinpox** [comment]
Cool, let me know if you need help testing!

JSON would be very nice indeed, as it would allow to feed the output in something that renders mustache templates. I'm specifically thinking of [mustache-go](https://github.com/cbroglie/mustache) here, which I already use to render [static content](https://pinpox.github.io/nixos/), e.g. for [documentation](https://github.com/pinpox/nixos/blob/9abc10d193969ed019a09a4b2866d26739c4b5cf/manual/manual.nix#L40-L45).

This would also allow to just create a dot or svg mustache template probably without further code.

**Comment by mickael-menu** [comment]
Sure, I'll keep you posted on this issue.

> I'm playing around with the list --format option, trying to output the necessary information as JSON. Looking at the docs, the only thing missing would be the ability to list the links of a note as a template value. Any ideas?

I also want to add a `json` option to the list `--format`. Doing it manually is a bit cumbersome and if you have quotes in your note titles they will break the JSON. Actually that might be a better start, as if we expose the `links` in `list` we could process a graph. Even though a dedicated graph JSON schema would be nicer to work with for `zk graph`.

I need to refactor some stuff in the models to have links more readily available. I expect that it could make the index retrieval slower for large ZKs, as it will need SQL joins to retrieve the links. Maybe adding an option `--with-links` could help for this.

**Comment by pinpox** [comment]
Good point, though I'd assume the ZKs would have to be *really* big it to be an issue performance-wise. Fully agree on the `json` option, that would be hugely useful and allow for great integration with a lot of other tools!

**Comment by mickael-menu** [comment]
I have a 30k notes ZK for testing and will see if it's an issue with it. If not, we can default with links and optionally add an option to disable links.

**Comment by mickael-menu** [comment]
I've been working on the JSON output for `zk list`: https://github.com/mickael-menu/zk/pull/52

I added a new `{{json x}}` template helper which outputs valid JSON values. We can rewrite your example to print JSON safely:

```sh
$ zk list --format '{ "title": {{json title}}, "tags": {{json tags}} }'
```
(The Handlebars template parser trips on `}}}`, so I added an extra space)

With the new `--header` and `--footer` options, you can output a valid JSON array:

```sh
$ zk list --format '{ "title": {{json title}}, "tags": {{json tags}} }' --delimiter , --header [ --footer ]
```

I built upon this to add two new formats `json` and `jsonl` ([JSON Lines](https://jsonlines.org/)) which are shortcuts for:

* `--format json` -> `--format "{{json .}}" --delimiter , --header [ --footer "]\n"`
* `--format jsonl` -> `--format "{{json .}}" --delimiter "\n" --footer "\n"`

**Comment by pinpox** [comment]
Very useful, thank you! 
Would it be possible to add an array of links as well? Having only the tags, a directed graph showing connections between notes is still not possible I think?

**Comment by mickael-menu** [comment]
No indeed. I actually added this today, but I'm not 100% sure I want to take this direction with `zk list`. Adding `links` gets quite hierarchical and feels a bit arbitrary (for example, why not have `backlinks` then?). Also I've pondered what information of the target should be part of the link. Probably its title and path but IMO it starts to feel redundant in the JSON.

Technically you should be able to post-process `zk list --format json` to retrieve the links yourself using `zk list --linked-by <source path> --format json`. This way you can get all the metadata you might need.

It's not a cop out though, I still see the benefit of having the links but I'm exploring the `zk graph` command instead. I think I will settle on the JSON format from D3.js as this lib is quite popular for these use cases and it is quite clean:

```json
{
    "nodes": [
        { "id": 1, "name": "A" },
        { "id": 2, "name": "B" }
    ],
    "links": [
        { "source": 1, "target": 2 }
    ]
}
```

I will just add some extra metadata in the objects. Do you have any idea what might be needed there?

[JGF](http://jsongraphformat.info/) is more specified, but at the same time I didn't see concrete applications using it.

**Comment by pinpox** [comment]
Just a thought: Maybe it would be wise to adhere to the `do one thing, do it right` principle here. I see more value in having a general JSON format representing the data, than some format specific to D3 or any other specific tool. You could declare the actual graphing as a non-feature and provide an example of how to use the external tools (like d3 or dot) to generate graphs. 

Adding this to each note:
> ```json
>     "links": [
>         "titleOrPath1", "titleOrPath2"
>     ]
> ```

in addition to the output you aready implemented above with the `--format json` flag, would be totally enough I think. What I have in mind would be not generating the graph's source from zk directly, but piping the json data to whatever tool I might want to use.


## Example

I'll add a full example, of what I personally would like to do. Take it with a grain of salt, this might be subjective, I'm just trying to give some ideas.

Assume the following notes:
<details>
  <summary>Note 1</summary>

  ```
---
title: testnote1
created: 2021-06-20
visibility: private
language: en
tags:
- testnote1
---

# testnote1

This is the first test note, [[testnote2]] is related to this.
  ```
</details>

<details>
  <summary>Note 2</summary>

  ```
---
title: testnote2
created: 2021-06-20
visibility: private
language: en
tags:
- testnote2
---

# testnote2

Here is a second test note. [[testnote3]] is related and [[testnote4]] aswell.
  ```
</details>
<details>
  <summary>Note 3</summary>

  ```
---
title: testnote3
created: 2021-06-20
visibility: private
language: en
tags:
- testnote3
---

# testnote3

The third note.
  ```
</details>
<details>
  <summary>Note 4</summary>

  ```
---
title: testnote4
created: 2021-06-20
visibility: private
language: en
tags:
- testnote4
---

# testnote4

Having four notes is enough for an example.

  ```
</details>


Running `zk list --format json --no-pager -m "testnote*" | jq` already gives most of the output I need, I have added the links array manually for each note for this example. Assume the command would print the following output:

<details>
  <summary>Expected output with `--format json`</summary>

  ```json
  [
  {
    "path": "testnote3.md",
    "title": "testnote3",
    "link": "[testnote3](testnote3)",
    "lead": "# testnote3",
    "body": "# testnote3\n\nThe third note.",
    "snippets": [
      "# testnote3\n\nThe third note."
    ],
    "rawContent": "edited/removed to keep example short",
    "wordCount": 18,
    "tags": [
      "testnote3"
    ],
    "metadata": {
      "created": "2021-06-20",
      "language": "en",
      "tags": [
        "testnote3"
      ],
      "title": "testnote3",
      "visibility": "private"
    },
    "created": "2021-06-20T14:14:45.031440861Z",
    "modified": "2021-06-20T14:14:55.598864042Z",
    "checksum": "5ae38fe0ff556bc37df7b9bfcce7893a0d090a6258c27f85dfaa441a23717fc1",
        "links": [ ]
  },
  {
    "path": "testnote4.md",
    "title": "testnote4",
    "link": "[testnote4](testnote4)",
    "lead": "# testnote4",
    "body": "# testnote4\n\nHaving four notes is enough for an example.",
    "snippets": [
      "# testnote4\n\nHaving four notes is enough for an example."
    ],
    "rawContent": "edited/removed to keep example short",
    "wordCount": 23,
    "tags": [
      "testnote4"
    ],
    "metadata": {
      "created": "2021-06-20",
      "language": "en",
      "tags": [
        "testnote4"
      ],
      "title": "testnote4",
      "visibility": "private"
    },
    "created": "2021-06-20T14:15:02.689644096Z",
    "modified": "2021-06-20T14:15:44.49852715Z",
    "checksum": "70dd52ebab26ca37f483c0a36277bd9121ebe4a5bcb8aad9ad5335915a904e91",
        "links": [ ]
  },
  {
    "path": "testnote1.md",
    "title": "testnote1",
    "link": "[testnote1](testnote1)",
    "lead": "# testnote1",
    "body": "# testnote1\n\nThis is the first test note, [[testnote2]] is related to this.",
    "snippets": [
      "# testnote1\n\nThis is the first test note, [[testnote2]] is related to this."
    ],
    "rawContent": "edited/removed to keep example short",
    "wordCount": 26,
    "tags": [
      "testnote1"
    ],
    "metadata": {
      "created": "2021-06-20",
      "language": "en",
      "tags": [
        "testnote1"
      ],
      "title": "testnote1",
      "visibility": "private"
    },
    "created": "2021-06-20T14:13:19.435952767Z",
    "modified": "2021-06-20T14:13:39.439388055Z",
    "checksum": "b216e9e1192701e4f9582a6964cbc8df93250977fd0d34ee410edaa557ecbc1f",
        "links": [ "testnote2" ]
  },
  {
    "path": "testnote2.md",
    "title": "testnote2",
    "link": "[testnote2](testnote2)",
    "lead": "# testnote2",
    "body": "# testnote2\n\nHere is a second test note. [[testnote3]] is related and [[testnote4]] aswell.",
    "snippets": [
      "# testnote2\n\nHere is a second test note. [[testnote3]] is related and [[testnote4]] aswell."
    ],
    "rawContent": "edited/removed to keep example short",
    "wordCount": 27,
    "tags": [
      "testnote2"
    ],
    "metadata": {
      "created": "2021-06-20",
      "language": "en",
      "tags": [
        "testnote2"
      ],
      "title": "testnote2",
      "visibility": "private"
    },
    "created": "2021-06-20T14:13:45.750153716Z",
    "modified": "2021-06-20T14:14:36.358996502Z",
    "checksum": "f3e54b0c4702663bc4e9f81c71567b04d5d1a4127c83eb2cf3a194299dc9c939",
        "links": [ "testnote3", "testnote4" ]
  }
]

  ```
</details>


Using any mustache parser (here I'm using [this simple go cli tool](https://github.com/cbroglie/mustache) ) it would allow to render a graph easily. 

I have saved the above to a file `notes.json` and this as `graph.mustache`

```dot
digraph D {

{{#.}}
"{{title}}" [shape=box, label="{{title}}" ]
{{#links}}
"{{title}}" -> "{{.}}" [color="black", style=dashed]
{{/links}}
{{/.}}
}
```

Running `cat notes.json | mustache graph.mustache | dot -Tpng > g.png` produces the following image:

![g](https://user-images.githubusercontent.com/1719781/122679069-ade54100-d1e9-11eb-8de6-541c58be2d50.png)

With a different template, any other graphing tool, e.g. `d3` could be "supported" in the same way, without requiring any development on `zk` itself.

# TL;DR

I would like to be able to do this:

```

zk list --format json --no-pager -m "testnote*" | mustache graph.mustache | dot
-Tpng > picture.png

```
And get the image above.  Sorry for the long comment, I hope you can find anything useful in it.

**Comment by mickael-menu** [comment]
Thanks for the food for thought, different use cases are always super helpful!

> Maybe it would be wise to adhere to the `do one thing, do it right` principle here. I see more value in having a general JSON format representing the data, than some format specific to D3 or any other tool.

I fully agree, that's why I wanted to find a standard spec to represent graphs. D3's format is quite clean but you convinced me to use a custom format more semantically related to the note domain, although I will keep the overall structure of D3.

```json
{
    "notes": [
        { "path": "a.md", "title": "A" },
        { "path": "b.md", "title": "B" }
    ],
    "links": [
        { "source": "a.md", "target": "b.md" }
    ]
}
``` 

If you look at [JGF](https://github.com/jsongraph/json-graph-specification/blob/master/examples/usual_suspects.json), [Sigma](http://sigmajs.org/) or even DOT, they all follow the same structure:

* A list of nodes/vertices with labels and IDs
* A list of edges/links connecting two IDs

Keeping this structure would make it trivial to convert to other formats with `jq` or a template engine like `mustache`.

> You could declare the actual graphing as a non-feature and provide an example of how to use external tools (like `d3` or `dot`) to generate graphs.

I think interconnectedness is an important part of a healthy notebook and I want to have links and graphs as first-class citizens in `zk`. I think using `zk list` with your example is not sufficient because:

* What about if you want to use different node labels than the note title?
* We are missing some interesting metadata which could make the graph more readable, such as:
    * Popover on hover to show the link snippet.
    * Colored arrows to show the link relationship (e.g. `proves`, `refutes`, `example`).
* It's not easy with the flat `list` structure to perform queries such as finding the backlinks of a given note or the shortest path between two notes.

Also having a dedicated `zk graph` command introduces a namespace for specific flags. For example, an option to include or not the outbound links to notes that are not matching the given filtering options.

**Comment by pinpox** [comment]
I see your point, having a better or more wide-used and common structure is definitely a plus!

> - What about if you want to use different node labels than the note title?
> - We are missing some interesting metadata which could make the graph more readable, such as:
> Popover on hover to show the link snippet.
> Colored arrows to show the link relationship (e.g. proves, refutes, example).

This is mostly why I put the links in the nodes itself in my example. Assuming the link only has a "target", the source is always the note itself, which would allow to access any field in it as part of the template. E.g. I used `title` in the example when doing this:
```mustache
{{#links}}
"{{title}}" -> "{{.}}" [color="black", style=dashed]
{{/links}}
```
But I could have used any other field. 

> It's not easy with the flat list structure to perform queries such as finding the backlinks of a given note or the shortest path between two notes.

I agree this could be an issue. I'll look into JGF and Sigma in more detail, I'm not very familiar with how to save graphs as json yet. That being said, there is always the option of using `jq`, which would be able to transform the data in any way needed. I think as long as it's performance is reasonable, theoretically any format that has the relevant data in it *somewhere* could be converted as needed with a `jq` query. 

I personally would not mind having to use `jq` to transform the data to my personal needs. It's obviously your decision what format you want to use in the end. I maybe should rephrase my feature request as "add a way to print out the data as json, with all necessary fields needed to create arbitrary graphs with it".

Thanks for looking into this by the way

**Comment by mickael-menu** [comment]
@pinpox I noticed you forked the project. If you want to test it out until I add `zk graph`, here's a patch of the changes I made to add `links` in `zk list`: [links.patch.txt](https://github.com/mickael-menu/zk/files/6704510/links.patch.txt)

**Comment by pinpox** [comment]
Hey, thanks for the patch! 
I'm currently using a nix derivation that build `zk` from a branch/repo of my choice to test it. You can find my package here, it will install zk from 0.5.0 from github https://github.com/pinpox/nixos/blob/main/packages/zk/default.nix

**Comment by mickael-menu** [comment]
@pinpox You can try out the [`feature/graph`](https://github.com/mickael-menu/zk/tree/feature/graph) branch. Still a work in progress but it should give you everything you need with:

```
zk graph --format json [filtering options as usual]
```

You can use the `path`, `sourcePath` and `targetPath` properties as node IDs.

Also I removed the `./go` script to build the project, you can use directly `make` or `make install` now. But it looks like you were not using it anyways in your `nix` package.

Please share any cool script you come up with to generate the actual graphs 👍

**Comment by pinpox** [comment]
Awesome! This makes it easy to generate graphs. Here is a working example:

File `template.mustache`:

```
digraph D {

{{#notes}}
"{{path}}" [shape=box, label="{{title}}" ]
{{/notes}}

{{#links}}
"{{sourcePath}}" -> "{{targetPath}}" [color="black", style=dashed]
{{/links}}
}
```

Then do:
```
zk graph --format json | mustache template.mustache | dot -Tpng > pic.png
```

You can also use `neato`, `fdp`, `sfdp`, `twopi` or `circo` instead of `dot` (see [here](https://graphviz.org/#roadmap) for details) to alter the resulting layout. 

Example using `sfdp` which seems  to look the best for my test data:
![sfdp](https://user-images.githubusercontent.com/1719781/123526839-cc13db00-d6da-11eb-82bf-a59cebffc5b2.png)
Obviously the graphs can be styled as you want, the template is just a minimal example

**Comment by mickael-menu** [comment]
Nice, thanks!

I think you should be able to adapt [the code of the link you shared](https://github.com/joashxu/zetteltools/blob/master/force_graph.html) easily. Just need to expand the `$data` variable inside after converting the JSON to the D3.js format with `jq`.

**Comment by pinpox** [comment]
Definitely! I'll try to get a working example with `d3`. One thing I noticed: there are "double links" when a note links multiple times to another. Is that intended/wanted? Of course it can be filtered with `jq` but maybe the links edges should be unique?

**Comment by mickael-menu** [comment]
> One thing I noticed: there are "double links" when a note links multiple times to another. Is that intended/wanted?

Yes because each link might have different associated metadata.

**Comment by pinpox** [comment]
I got a first working example with d3:
![image](https://user-images.githubusercontent.com/1719781/123978519-37ea9200-d9c0-11eb-8213-35147fd0d1c7.png)
I still want to improve it a bit and will then upload it to some public repository, but if you want to take a look in the meantime the files used are here:
https://gist.github.com/pinpox/17496ff61ad4012dccb238d8f3deca81

**Comment by mickael-menu** [comment]
Nice, thanks! Chrome gave me a CORS error but I could test it easily running `python3 -m http.server` in the folder.

On neuron they have been working on graphs with D3 as well: https://github.com/srid/neuron/issues/589

We probably need some fine-tuning for the attraction settings. Or maybe using a much smaller font and use zooming.
<img width="1069" alt="Screenshot 2021-07-01 at 11 32 40" src="https://user-images.githubusercontent.com/58686775/124101839-0de8ac80-da60-11eb-8539-f66a39fae260.png">

**Comment by pinpox** [comment]
Yes, the attraction simulation is not optimal. Specifically this example would be a better fit:

> When using D3’s force layout with a disjoint graph, you typically want the positioning forces (d3.forceX and d3.forceY) instead of the centering force (d3.forceCenter). The positioning forces, unlike the centering force, prevent detached subgraphs from escaping the viewport.

(Taken from this example: https://observablehq.com/@d3/disjoint-force-directed-graph?collection=@d3/d3-force )

I have tried modifying my code to use the example settings:
```js
  const simulation = d3.forceSimulation(nodes)
      .force("link", d3.forceLink(links).id(d => d.id))
      .force("charge", d3.forceManyBody())
      .force("x", d3.forceX())
      .force("y", d3.forceY());

```
But can't get it to work. Seems like the different d3 versions don't work exactly the same way, but I havn't been able to fix it. Using the example above should result in the nodes been stretched apart into a circle when disconnected. 

There is [another example here](https://observablehq.com/@brunolaranjeira/d3-v6-force-directed-graph-with-directional-straight-arrow) that also looks way better, but I have the same problem. 
If you know how to use thee `forceX()` and `forceY()` instead of the forces used in my example, please let me know. I'll keep trying and report back if I get it to work.

**Comment by mickael-menu** [comment]
I actually never worked with D3 before but I've seen it being used a lot for this use case. If you're able to make the font size much smaller that could be a good workaround.

**Comment by pinpox** [comment]
Small update. Got the correct forces working. The labels are still messy and overlapping, but I think I'm on the right track now. The simulation now prevents disconnected nodes to "fly away", they all nicely arrange into a circle. I'll either add collision to the text labels too or use some placement algorithm, there seem to be a few options. Another, simpler, way might be just making the links longer or the text smaller. 

![image](https://user-images.githubusercontent.com/1719781/124154889-34293f00-da96-11eb-9988-399a27396516.png)


Here is the [updated source](https://gist.github.com/pinpox/bbb876f8643a399205e0a5eb47c4a321)

**Comment by pinpox** [comment]
@mickael-menu Is https://github.com/mickael-menu/zk/releases/tag/v0.6.0 compatible with the command above? The `graph` subcommand seems to be missing?

**Comment by mickael-menu** [comment]
Not yet, the graph feature is a significant change and I need more time to clean up the code. I merged `main` in `feature/graph` so it should be up to date with the recent changes.

**Comment by mickael-menu** [comment]
@pinpox I merged the feature in `main`. Let me know if everything is fine and I'll do a proper release.

You might need to reset your database if you used the old `feature/graph` branch with `rm .zk/notebook.db`.

**Comment by pinpox** [comment]
@mickael-menu Looks good! I build from master and the command seems to produce correct JSON. 

Maybe it would be a good idea to document an example of how to render a proper graph with it somewhere in the docs?

**Comment by mickael-menu** [comment]
That's a good idea. I didn't spend much time playing with this though, would you consider writing a Markdown page dedicated to the `graph` command?

**Comment by almereyda** [comment]
As a late addition to the reference of the non-working `zetteltools` in the OP, the following adjustments allowed me to render the graph of links between the notes in my Zettelkasten with it.

<details>

<summary>The Zettelkasten is a view on a Logseq Journal.</summary>

```shell
$ cat graph.sh                                               
#!/usr/bin/env bash
set -m

# https://unix.stackexchange.com/questions/50179/what-happens-when-you-delete-a-hard-link
rm -rf ./zk/*.md

# https://unix.stackexchange.com/questions/40634/in-ubuntu-is-there-a-way-to-virtually-merge-two-folders-without-unionfs-or-aufs
cp -anl "$PWD/journals/"* zk/
cp -anl "$PWD/pages/"* zk/

zk -W zk graph --format json > graph/notes.json

http -q graph & xdg-open http://localhost:8000 && fg
```

[`http`](https://crates.io/crates/https) is a small webserver that works similarily to `python -m http.server`.

</details>

In the `graph/` subdirectory we find [a forked revision of `index.html` and `graph.js`](https://gist.github.com/almereyda/4bebc52d1e45a22ff73bc07a4927de22/revisions) supplied by the pinpox gist above.

**Comment by pinpox** [comment]
@almereyda I'm struggeling a bit to understand. Did you use the gists, your forked zetteltools repo or both?

Could you provide a complete minimal working example?
u 

> That's a good idea. I didn't spend much time playing with this though, would you consider writing a Markdown page dedicated to the `graph` command?

@mickael-menu I'd like to include a full working example in the documentation with the above if possible

**Comment by mickael-menu** [comment]
> @mickael-menu I'd like to include a full working example in the documentation with the above if possible. The forces in `zetteltools` seem to make more sense, at least the example image in the upsteam repo looks better.

Sure, feel free to open a PR on the repo 👍

**Comment by TSoli** [comment]
> @almereyda I'm struggeling a bit to understand. Did you use the gists, your forked zetteltools repo or both? Could you provide a complete minimal working example? I cloned your fork and running `python3 zettvis.py ~/Notes/` gives no links and no titles on hover:
> 
> ![image](https://user-images.githubusercontent.com/1719781/142198969-4ac23116-5179-4cbc-aa1b-09c78f139ea2.png)
> 
> > That's a good idea. I didn't spend much time playing with this though, would you consider writing a Markdown page dedicated to the `graph` command?
> 
> @mickael-menu I'd like to include a full working example in the documentation with the above if possible. The forces in `zetteltools` seem to make more sense, at least the example image in the upsteam repo looks better.

I have the same issue. With your gist it also seems not all the notes fit in the frame and I am not sure how to zoom. Did anyone figure out how to get the `zetteltools` version working or a nicer graph with zoom/pan?

**Comment by kha-dinh** [comment]
I am making a graph visualization web app with NextJS for those who are still looking. 
it's highly experimental (and buggy), but you can check it out at
https://github.com/kha-dinh/grzk/

**Comment by WhyNotHugo** [comment]
None of the above commands using `mustache` worked for me, `zk graph --format json | mustache template.mustache` fails with:

```
[note data here...] not a serializable argument for {{json}}: json: error calling MarshalJSON for type *core.lazyStringer: invalid character 't' after top-level value


Found 915 notes
Error: yaml: line 117: did not find expected node content
```

The output format changed substantially since those comments, right? Any ideas how to pipe `zk` into `dot` with the current format?

**Comment by WhyNotHugo** [comment]
Oh, actually, it's `zk graph --format json > /dev/null` which fails to serialise:

```
 94 [] map[] 2025-01-11 21:51:53.108203539 +0000 UTC 2025-01-09 22:25:00.634903118 +0000 UTC 2a746e4b4f5a903cc9c57e6e6ddcbc88a02256c6095afdf46534ff04743af0ec map[CARGO_HOME:/home/hugo/.local/state/cargo CHARSET:UTF-8 COLORTERM:truecolor DBUS_SESSION_BUS_ADDRESS:unix:path=/run/user-1000/bus DISPLAY::0 DOCKER_CONFIG:/home/hugo/.config/docker DOCKER_HOST:unix:///run/user-1000/docker.sock EDITOR:nvim FZF_CTRL_T_COMMAND:fd --strip-cwd-prefix FZF_DEFAULT_COMMAND:ag -g . GNUPGHOME:/home/hugo/.local/share/gnupg GOPATH:/home/hugo/.cache/golang GOPROXY:direct HOME:/home/hugo I3SOCK:/run/user-1000/sway-ipc.1000.3552.sock IPYTHONDIR:/home/hugo/.local/share/ipython KEYTIMEOUT:1 LANG:en_GB.UTF-8 LC_COLLATE:C LC_MONETARY:nl_NL.UTF-8 LC_NUMERIC:en_GB.UTF-8 LC_TIME:en_DK.UTF-8 LESS:-RX --mouse --quit-if-one-screen LESSHISTFILE:/home/hugo/.local/share/less_history LOGNAME:hugo MANPAGER:less -R --use-color -Dd+r -Du+b MANPATH:/usr/share/man:/usr/local/share/man MANROFFOPT:-P -c MANWIDTH:92 MUSL_LOCPATH:/usr/share/i18n/locales/musl NOTMUCH_CONFIG:/home/hugo/.config/notmuch/config OLDPWD:/home/hugo/tmp PAGER:bat -p PATH:/home/hugo/.local/bin:/home/hugo/.local/state/cargo/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin PIP_REQUIRE_VIRTUALENV:true PSQL_HISTORY:/home/hugo/.local/share/psql_history PWD:/home/hugo/tmp/bbb876f8643a399205e0a5eb47c4a321 QT_AUTO_SCREEN_SCALE_FACTOR:1 QT_QPA_PLATFORM:wayland QT_SCALE_FACTOR:1 QT_WAYLAND_DISABLE_WINDOWDECORATION:1 RUSTUP_HOME:/home/hugo/.local/state/rustup SHELL:/bin/zsh SHLVL:3 SSH_AUTH_SOCK:/run/user-1000/hissh-agent STARSHIP_SESSION_KEY:4490233093104776 STARSHIP_SHELL:zsh SWAYSOCK:/run/user-1000/sway-ipc.1000.3552.sock TERM:foot TERMINAL:foot TZDIR:/etc/zoneinfo USER:hugo VIRTUALENVWRAPPER_SCRIPT:/usr/bin/virtualenvwrapper.sh VISUAL:nvim WAYLAND_DISPLAY:wayland-1 WINEDLLOVERRIDES:winemenubuilder.exe=d WORKON_HOME:/home/hugo/.local/state/virtualenvs XCURSOR_SIZE:24 XCURSOR_THEME:Adwaita XDG_CACHE_HOME:/home/hugo/.cache XDG_CONFIG_HOME:/home/hugo/.config XDG_DATA_DIRS:/home/hugo/.local/share/flatpak/exports/share:/var/lib/flatpak/exports/share:/usr/local/share:/usr/share XDG_DATA_HOME:/home/hugo/.local/share XDG_RUNTIME_DIR:/run/user-1000 XDG_SEAT:seat0 XDG_SESSION_TYPE:wayland XDG_STATE_HOME:/home/hugo/.local/state XDG_VTNR:1 ZK_NOTEBOOK_DIR:/home/hugo/zk ZK_SHELL: _:/usr/bin/zk _JAVA_AWT_WM_NONREPARENTING:1 _JAVA_OPTIONS:-Dswing.aatext=true _VIRTUALENVWRAPPER_API: mkvirtualenv rmvirtualenv lsvirtualenv showvirtualenv workon add2virtualenv cdsitepackages cdvirtualenv lssitepackages toggleglobalsitepackages cpvirtualenv setvirtualenvproject mkproject cdproject mktmpenv wipeenv allvirtualenv npm_config_cache:/home/hugo/.cache/npm]}: not a serializable argument for {{json}}: json: error calling MarshalJSON for type *core.lazyStringer: invalid character 't' after top-level value


Found 915 notes
```

Maybe this is a bug and I should open a new issue?

**Comment by tjex** [comment]
That looks like a bug that we fixed in #440 to do with escaping quotation marks in note titles.

Are you on the latest version? 0.14.1

**Comment by WhyNotHugo** [comment]
I'm using zk-0.14.1-r4, from Alpine packages.

**Comment by tjex** [comment]
Hmm. Odd. It could be that something else is breaking the json formatting.
I can run the command without any problems.

Would be worth making a dummy notebook with a few random notes to try the command there in isolation.

**Comment by WhyNotHugo** [comment]
Created an issue with a reproduction example at https://github.com/zk-org/zk/issues/492
